doc/BACKUP.md: Document backup strategies for `lightningd`. (#4207)
ElementsProject/lightning: doc/BACKUP.md

of the build directory.

    lightning-hsmtool generatehsm hsm_secret

Then enter the BIP39 words, plus an optional passphrase.

You can regenerate the same `hsm_secret` file using the same BIP39
words, which again, you can back up on paper.

Recovery of the `hsm_secret` is sufficient to recover any onchain
funds.
Recovery of the `hsm_secret` is necessary, but insufficient, to recover
any in-channel funds.
To recover in-channel funds, you need to use one or more of the other
backup strategies below.

Database File Backups
---------------------

This is the least desirable backup strategy, as it *can* lead to loss
of all in-channel funds if you use it.
However, having *no* backup strategy at all *will* lead to loss of all
in-channel funds, so this is still better than nothing.

While `lightningd` is not running, just copy the `lightningd.sqlite3` file
in the `$LIGHTNINGDIR` on backup media somewhere.

To recover, just copy the backed up `lightningd.sqlite3` into your new
`$LIGHTNINGDIR` together with the `hsm_secret`.

You can also use any automated backup system as long as it includes the
`lightningd.sqlite3` file (and optionally `hsm_secret`, but note that

> as a secret key, hackers getting a copy of your backups may allow them

Reviewer: hackers? Not sure we should jump on the train calling thieves hackers.

http://www.catb.org/jargon/html/index.html

channels:

* Channels with peers that do not support `option_dataloss_protect`.
  * Most nodes on the network already support `option_dataloss_protect`
    as of November 2020.
  * If the peer does not support this option, then the entire channel
    funds will be revoked by the peer.
  * Specially-coded peers can pretend to support this option, but end
    up taking the entire channel funds (or otherwise more than what it
    should have) when you try to recover.
    Standard releases of Lightning Network node software do not do this
    (they just support this option according to the standard) but note
    that anyone can modify their node software, especially with our
    well-documented open-source code.
* Channels created *after* the copy was made are not recoverable.
  * Data for those channels does not exist in the backup, so your node
    cannot recover them.

Because of the above, this strategy is discouraged: you *can* potentially
lose all funds in open channels.

However, again, note that a "no backups #reckless" strategy leads to
*definite* loss of funds, so you should still prefer this strategy rather

> than not have **any** backups at all.

Reviewer: Emphasis on word **any** is a bit misleading. There is no comparision (with other emphasized word nearby). I would prefer something like "so you should still prefer this strategy rather than having no backups at all."

Author: "you should still prefer **this** strategy rather than have **no** backups at all." reads better?

Reviewer: Sure. That works well.

Even if you have the better options below, you might still want to do

> this as a fallback, as long as you ensure:

Reviewer: as long as you: (and see below)

> * Attempt to recover using the other backup options below first.

Reviewer: This is good as it is. Actually it does not go well when there is the "ensure" word above.

  Any one of them will be better than this backup option.

> * Recovering by this method ***MUST*** always be the ***last*** resort.

Reviewer: Change to something like "Use this method **ONLY** as a ***last*** resort" (goes well with "as long as you:" above)

Author: Went with "Recover by this method **ONLY** as a ***last*** resort"

on a recovery disk or similar, then:

* Connect the surviving device.
* Mount the partition/USB flash disk in `degraded` mode:
  * `mount -o degraded /dev/sdXX /mnt/point`
* Copy the `lightningd.sqlite3` and `hsm_secret` to new media.
  * Do **not** write to the degraded `btrfs` mount!
* Start up a `lightningd` using the `hsm_secret` and `lightningd.sqlite3`
  and close all channels and move all funds to onchain cold storage you
  control, then set up a new Lightning node.

If the device that fails is the USB flash disk, you can replace it using
BTRFS commands.
You should probably stop your Lightning node while doing this.

* `btrfs replace start /dev/sdOLD /dev/sdNEW ${BTRFSMOUNTPOINT}`.
* Monitor status with `btrfs replace status ${BTRFSMOUNTPOINT}`.

More sophisticated setups with more than two devices are possible.
Take note that "RAID 1" in `btrfs` means "data is copied on up to two
devices", meaning only up to one device can fail.
You may be interested in `raid1c3` and `raid1c4` modes if you have
three or four storage devices.
BTRFS would probably work better if you were purchasing an entire set
of new storage devices to set up a new node.

> On BSD you can use a ZFS RAID-Z setup, which is probably better than BTRFS

Reviewer: The general advice (like first paragraph in this section also) could be moved together, BTRFS could be turned into its own sub-section so that it is optically skippable (sp?).

BTW this review comes from a point of view of a non-native speaker. So take it with a grain of salt.

Author: Okay.  I guess we can list software methods of RAID-1, such as:

* `mdadm`
* BTRFS
* ZFS

  This creates an initial copy of the database at the NFS mount.
* Add these settings to your `lightningd` configuration:
  * `important-plugin=/path/to/backup.py`
  * `backup-destination=file:///path/to/nfs/mount`
* Restart your Lightning node.

It is recommended that you use a network-mounted filesystem for the backup
destination.

Alternately, you *could* put it in another storage device (e.g. USB flash
disk) on the same physical location, but if so, you might want to use the
above `btrfs` RAID-1 instead, as that tends to be more performant
(though using this `backup` plugin does not require you to repartition an
existing hard disk, which might matter to you if you are using an
everyday-use laptop or desktop).

To recover:

* Re-download the `backup` plugin and install Python 3 and the
  requirements of `backup`.
* `/path/to/backup-cli restore file:///path/to/nfs/mount ${LIGHTNINGDIR}`

If your backup destination is a network-mounted filesystem that is in a
remote location, then even loss of all hardware in one location will allow
you to still recover your Lightning funds.

However, if instead you are just replicating the database on another

> storage device on a single location, you remain vulnerable to disasters

Reviewer: in a single location?

Author: Right, "in" is better than "on" here.

* Attempt to recover using the other backup options below first.
  Any one of them will be better than this backup option.
* Recover by this method **ONLY** as a ***last*** resort.
* Recover using the most recent backup you can find.
  Take time to look for the most recent available backup.

Again, this strategy can lead to only ***partial*** recovery of funds,
or even to complete failure to recover, so use the other methods below
first to recover!

### Backing Up While `lightningd` Is Running

Since `sqlite3` will be writing to the file while `lightningd` is running,
`cp`ing the `lightningd.sqlite3` file while `lightningd` is running may
result in the file not being copied properly if `sqlite3` happens to be
committing database transactions at that time, potentially leading to a
corrupted backup file that cannot be recovered from.

Instead, the **proper** way to do this would be:

* `echo "VACUUM INTO 'backup.sqlite3';" | sqlite3 lightningd.sqlite3`
* `sync`
* `mv backup.sqlite3 ${YOURBACKUPLOCATION}/lightningd.sqlite3`

This creates a consistent snapshot of the database, sampled in a
transaction, that is assured to be openable later by `sqlite3`.

> The operation of the `lightningd` process will be paused while `sqlite3`

Author: ***THIS IS FACTUALLY INCORRECT***.

I did a stress test where a program continuously makes many small transactions to update a database.  When you do something like `.backup 'backup.sqlite3'` or `VACUUM INTO `backup.sqlite3';` in a *separate* `sqlite3`, then either the main program gets an `SQLITE_BUSY`, ***or*** the separate `sqlite3` returns a "Database is locked" error.

I suspect that an `SQLITE_BUSY` will cause `lightningd` to crash, so this is probably strongly not recommended to use a separate process to back up.

We can do:

* Expose an `sqlite3_backup` command which performs the `VACUUM INTO` query inside `lightningd`.
* Call into `sqlite3_busy_timeout` with a "reasonable" large value, say 5000 (5 seconds). (Equivalently, call `PRAGMA busy_timeout = 5000;` query at the same time we do `PRAGMA foreign_keys=on;`).  Then, even if the `.backup` or `VACUUM INTO` takes up to 5 seconds to back up, `lightningd` will just wait.

The latter solution is easier but there is always the possibility that the backup process will take more than whatever timeout we select.  The earlier solution is more reliable and there is no timeout, but note that if say the target location is slow (e.g. an NFS mount) then `lightningd` will suspend indefinitely.

Thoughts?

Reviewer: Thank you @ZmnSCPxj for ideas and investigating possible issues with running backup in separate `sqlite3` process!

Exposing the `sqlite3_backup` command would be great! It sounds like a very nice and clear way to do backup.

Notice that you need to ensure that the `hsm_secret` is only readable by
the user, and is not writable, as otherwise `lightningd` will refuse to
start.
Hence the `chmod 0400 hsm_secret` command.

Alternately, if you are deploying a new node that has no funds and
channels yet, you can generate BIP39 words using any process, and
create the `hsm_secret` using the `hsmtool generatehsm` command.
If you did `make install` then `hsmtool` is installed as
`lightning-hsmtool`, else you can find it in the `tools/` directory
of the build directory.

    lightning-hsmtool generatehsm hsm_secret

Then enter the BIP39 words, plus an optional passphrase.

You can regenerate the same `hsm_secret` file using the same BIP39
words, which again, you can back up on paper.

Recovery of the `hsm_secret` is sufficient to recover any onchain
funds.
Recovery of the `hsm_secret` is necessary, but insufficient, to recover
any in-channel funds.
To recover in-channel funds, you need to use one or more of the other
backup strategies below.

Database File Backups
---------------------

This is the least desirable backup strategy, as it *can* lead to loss
of all in-channel funds if you use it.

> However, having *no* backup strategy at all *will* lead to loss of all

Reviewer 2: Not sure it is true, as eventually the other user will force-close the channels and you will get the BTC back to your address? right?

Author: If the peer does not have `option_static_remotekey` then the address it goes to is not on the path derived from `hsm_secret`, but requires the derivation keys stored in the db.  Though lots of nodes today already support `option_static_remotekey`.

Author: You also need to wait for the other side to actually force close, and that could take a while.

Reviewer 3: No, if a channel was opened without the static key option the payout address is dependent on the last state (address is tweaked with the channel state), and the wallet will not be able to recognize the address as its own.

Worse: relying on the peer to do the right thing might expose you to fraud if they decide they prefer an older state where they had a bigger share of the channel capacity.

to steal your funds, even in-channel funds).

This backup method is undesirable, since it cannot recover the following
channels:

* Channels with peers that do not support `option_dataloss_protect`.
  * Most nodes on the network already support `option_dataloss_protect`
    as of November 2020.
  * If the peer does not support this option, then the entire channel
    funds will be revoked by the peer.
  * Specially-coded peers can pretend to support this option, but end
    up taking the entire channel funds (or otherwise more than what it
    should have) when you try to recover.
    Standard releases of Lightning Network node software do not do this
    (they just support this option according to the standard) but note
    that anyone can modify their node software, especially with our
    well-documented open-source code.
* Channels created *after* the copy was made are not recoverable.
  * Data for those channels does not exist in the backup, so your node
    cannot recover them.

Because of the above, this strategy is discouraged: you *can* potentially
lose all funds in open channels.

However, again, note that a "no backups #reckless" strategy leads to

> *definite* loss of funds, so you should still prefer *this* strategy rather

Reviewer 2: again, what if the other user close the channels? wouldn't you get the money back to your wallet?

Reviewer 3: Just to add a bit more complexity: if we use `option_upfront_shutdown_script` we may actually get the funds back to our wallet if the peer closes, because that removes the tweak to the their_unilateral/to_us transaction.

  requirements of `backup`.
* `/path/to/backup-cli restore file:///path/to/nfs/mount ${LIGHTNINGDIR}`

If your backup destination is a network-mounted filesystem that is in a
remote location, then even loss of all hardware in one location will allow
you to still recover your Lightning funds.

However, if instead you are just replicating the database on another
storage device in a single location, you remain vulnerable to disasters
like fire or computer confiscation.

PostgreSQL Cluster
------------------

`lightningd` may also be compiled with PostgreSQL support.
PostgreSQL is generally faster than SQLITE3, and also supports running a
PostgreSQL cluster to be used by `lightningd`, with automatic replication
and turnover in case an entire node of the PostgreSQL cluster fails.

Setting this up, however, is more involved.

By default, `lightningd` compiles without PostgreSQL support.
To enable it, you have to install a developer version of `libpq`.
On most Debian-derived systems that would be `libpq-dev`.
To verify you have it properly installed on your system, check if the
following command gives you a path:

    pg_config --includedir

> 

Reviewer 4: I didn't know it. When I built it, everything worked out of the box. Pretty useful to know. 
I probably had already the library installed.

Author: Okay, maybe the Lightning+PostgreSQL can be a separate PR?

Reviewer 4: The document you mean?
Something like `doc/postgresQL.md`?

Anyway This part in the backup strategies seems pretty right to me. At least to mention a more extended part in the doc directory.  @fiatjaf Has thought about that?

Author: Yes, the document I mean.

Versioning may also matter to you.
For example, Debian Stable ("buster") as of late 2020 provides PostgreSQL 11.9
for the `libpq-dev` package, but Ubuntu LTS ("focal") of 2020 provides
PostgreSQL 12.5.
Debian Testing ("bullseye") uses PostgreSQL 13.0 as of this writing.
PostgreSQL 12 had a non-trivial change in the way the restore operation is
done for replication.

> You should use the same PostgreSQL version of `libpq-dev` as what you run

Reviewer 4: I encountered the problem myself. I would add to always read the official synchronous replication PostgresQL guide for the **specific version** you are using.

Author: Okay, I will add that.

Reviewer 3: Agreed, there is little point in replicating the official doc here, it may just end up with stale information, requiring updates from us. If there is an authoritative source, link to it.

on your cluster, which probably means running the same distribution on
your cluster.

> 

Reviewer 4: Personally I have added on the cluster machines the Postgres .deb repository in this way I can keep the versions aligned.
I didn't encounter any problem but maybe it is not wise to exit from the Postgres version installed with the Debian distribution. What you guys think?

Author: I would have used the same distro throughout just to reduce inter-version problems, which is why I suggested the above.  Mixing deb repositories has gotten me into trouble before when I wanted to upgrade my OS (but upgrading an OS is always fraught, sometimes it is best to just have two small OS partitions and alternate installing OS's between them rather than upgrading an existing install...)

You then have to compile `lightningd` with PostgreSQL support.

* Clone or untar a new source tree for `lightning` and `cd` into it.
  * You *could* just use `make clean` on an existing one, but for the
    avoidance of doubt (and potential bugs in our `Makefile` cleanup rules),
    just create a fresh source tree.
* `./configure`
  * Add any options to `configure` that you normally use as well.
* Double-check the `config.vars` file contains `HAVE_POSTGRES=1`.
  * `grep 'HAVE_POSTGRES' config.vars`
* `make`
* If you install `lightningd`, `sudo make install`.

If you were not using PostgreSQL before but have compiled and used
`lightningd` on your system, the resulting `lightningd` will still
continue supporting and using your current SQLITE3 database;
it just gains the option to use a PostgreSQL database as well.

To set up a cluster for a brand new node, follow this (external)
[guide by @gabridome][gabridomeguide].

[gabridomeguide]: https://github.com/gabridome/docs/blob/master/c-lightning_with_postgresql_reliability.md

> 

Reviewer 4: The guide is not maintained as the versions of Postgres evolve. Please always check also the Postgres Guide about synchronous replication for your specific version.

The above guide assumes you are setting up a new node from scratch.
If you want to continue a node that started using an SQLITE3 database,
use the [tool by @fiatjaf][fiatjafmcldsp] to convert your SQLITE3 database
to a PostgreSQL database.
Follow up to just *before* "Connect to the database" on the @gabridome
guide, then use the @fiatjaf `mcldsp` tool with the `postgres` specification
you generated, then continue with the rest of the @gabridome guide
(though you should probably do some more double-checking and tire-kicking
in the "Connect to the database" stage you resume at, such as checking if
`listpeers` still lists the same channels as you had, and so on).

> 

Reviewer 4: Wise suggestion IMO.

Backing Up Your C-Lightning Node
================================

Lightning Network channels get their scalability and privacy benefits
from the very simple technique of *not telling anyone else about your
in-channel activity*.
This is in contrast to onchain payments, where you have to tell everyone
about each and every payment and have it recorded on the blockchain,

> leading to scaling problems (you have to push data at everyone) and

Reviewer 3: ```suggestion
leading to scaling problems (you have to push data to everyone, everyone needs to verify every transaction) and
```

privacy problems (everyone knows every payment you were ever involved in).

Unfortunately, this removes a property that onchain users are so used
to, they react in surprise when learning about this removal.
Your onchain activity is recorded in all archival fullnodes, so if you
forget all your onchain activity because your storage got fried, you
just go redownload the activity from the nearest archival fullnode.

But in Lightning, since *you* are the only one storing all your
financial information, you ***cannot*** recover this financial

> information anywhere else.

Reviewer 3: ```suggestion
information from anywhere else.
```

> This means that on Lightning, you **have to** responsibly back up your

Reviewer 3: ```suggestion
This means that on Lightning, **you have to** responsibly back up your
```
:wink:

financial information yourself, using various processes and automation.

The discussion below assumes that you know where you put your
`$LIGHTNINGDIR`, and you know the directory structure within.
By default your `$LIGHTNINGDIR` will be in `~/.lightning/${COIN}`.
For example, if you are running `--mainnet`, it will be
`~/.lightning/bitcoin`.

`hsm_secret`

> ------------

Reviewer 3: Mix of `----` and `###` headings makes it more difficult to judge the overall structure. Please use `###` prefixes wherever possible.

Author: I mildly dislike the `###` prefixes since to me "more `#` = more heavier" i.e. `###` should map to `<h1>`, `#` should map to `<h3>`.  *shrug*.  Changed anyway.

You need a copy of the `hsm_secret` file regardless of whatever backup
strategy you use.
This is just 32 bytes, and you can do something like the below and

> write the hexadecimal digits a few times on a piece of paper:

Reviewer 3: Maybe mention that this file does not change after its creation, so backing it up when first setting up the node is sufficient and no ongoing backup is required, in contrast to the DB which requires continuous backups.

    10: e1f3 1d45 d904 823c 77b7 1e18 fd93 1676
    HEX
    xxd -r hsm_secret_hex.txt > hsm_secret
    chmod 0400 hsm_secret

Notice that you need to ensure that the `hsm_secret` is only readable by
the user, and is not writable, as otherwise `lightningd` will refuse to
start.
Hence the `chmod 0400 hsm_secret` command.

Alternately, if you are deploying a new node that has no funds and
channels yet, you can generate BIP39 words using any process, and
create the `hsm_secret` using the `hsmtool generatehsm` command.
If you did `make install` then `hsmtool` is installed as
`lightning-hsmtool`, else you can find it in the `tools/` directory
of the build directory.

    lightning-hsmtool generatehsm hsm_secret

Then enter the BIP39 words, plus an optional passphrase.

You can regenerate the same `hsm_secret` file using the same BIP39
words, which again, you can back up on paper.

Recovery of the `hsm_secret` is sufficient to recover any onchain
funds.
Recovery of the `hsm_secret` is necessary, but insufficient, to recover
any in-channel funds.
To recover in-channel funds, you need to use one or more of the other
backup strategies below.

Database File Backups

> ---------------------

Reviewer 3: IMHO the methods should be sorted in descending order of desirability, putting this type first gives absolutely the wrong impression.

This is the least desirable backup strategy, as it *can* lead to loss
of all in-channel funds if you use it.
However, having *no* backup strategy at all *will* lead to loss of all
in-channel funds, so this is still better than nothing.

While `lightningd` is not running, just copy the `lightningd.sqlite3` file
in the `$LIGHTNINGDIR` on backup media somewhere.

To recover, just copy the backed up `lightningd.sqlite3` into your new
`$LIGHTNINGDIR` together with the `hsm_secret`.

You can also use any automated backup system as long as it includes the
`lightningd.sqlite3` file (and optionally `hsm_secret`, but note that
as a secret key, thieves getting a copy of your backups may allow them
to steal your funds, even in-channel funds).

This backup method is undesirable, since it cannot recover the following
channels:

* Channels with peers that do not support `option_dataloss_protect`.
  * Most nodes on the network already support `option_dataloss_protect`
    as of November 2020.
  * If the peer does not support this option, then the entire channel

>     funds will be revoked by the peer.

Reviewer 3: Technically not true: the peer may be well-behaved, and receive back his share of the funds, but the user's funds will be lost forever into an unrecoverable address, due to the tweak.

Even if the peer is not well-behaved, it can at most grab the funds that belonged to it some time in the past, not all funds. The peer would then go back to that point in time, grab it's now bigger share, and still sinkhole the user's share.

Author: How about "up to the entire channel funds can be revoked by the peer"?

Author: No, wait, this is true if the `option_dataloss_protect` is *not* supported, which is what qualifies the statement.  Without `option_dataloss_protect` we would get confused with the `reestablish_connection` and try to broadcast our obsolete revoked commitment transaction, and the peer will react by revoking the revoked commitment transaction.

  * Specially-coded peers can pretend to support this option, but end
    up taking the entire channel funds (or otherwise more than what it
    should have) when you try to recover.
    Standard releases of Lightning Network node software do not do this
    (they just support this option according to the standard) but note
    that anyone can modify their node software, especially with our

>     well-documented open-source code.

Reviewer 3: Doesn't change the behavior above: the peer can prevent the user from recovering its share of the funds by not sharing the commitment point.

Reviewer 3: Generally speaking I'd reduce the level of detail here slightly, to "if the peer is non-cooperative you won't be able to recover your funds".

* Channels created *after* the copy was made are not recoverable.
  * Data for those channels does not exist in the backup, so your node
    cannot recover them.

Because of the above, this strategy is discouraged: you *can* potentially
lose all funds in open channels.

However, again, note that a "no backups #reckless" strategy leads to
*definite* loss of funds, so you should still prefer *this* strategy rather
than having *no* backups at all.

Even if you have the better options below, you might still want to do
this as a worst-case fallback, as long as you:

* Attempt to recover using the other backup options below first.
  Any one of them will be better than this backup option.
* Recover by this method **ONLY** as a ***last*** resort.
* Recover using the most recent backup you can find.
  Take time to look for the most recent available backup.

Again, this strategy can lead to only ***partial*** recovery of funds,
or even to complete failure to recover, so use the other methods below
first to recover!

> ### Backing Up While `lightningd` Is Running

Reviewer 3: Ok, while the above should be the last option presented, this one should be even laster! :wink:

corrupted and you should make another copy.

In order to make a proper uncorrupted backup of the SQLITE3 file
while `lightningd` is running, we would need to have `lightningd`
perform the backup itself, which, as of the version at the time of
this writing, is not yet implemented.

Even if the backup is not corrupted, take note that this backup
strategy should still be a last resort; recovery of all funds is
still not assured with this backup strategy.

Filesystem Redundancy
---------------------

You can set up a RAID-1 with multiple storage devices, and point the
`$LIGHTNINGDIR` to the RAID-1 setup.
That way, failure of one storage device will still let you recover
funds.

You can use a hardware RAID-1 setup, or just buy multiple commodity
storage media you can add to your machine and use a software RAID,
such as (not an exhaustive list!):

* `mdadm` to create a virtual volume which is the RAID combination
  of multiple physical media.
* BTRFS RAID-1 or RAID-10, a filesystem built into Linux.
* ZFS RAID-Z, a filesystem that cannot be legally distributed with the Linux
  kernel, but can be distributed in a BSD system, and can be installed 
  on Linux with some extra effort, see

>   [ZFSonLinux](https:://zfsonlinux.org).

Reviewer 3: ```suggestion
  [ZFSonLinux](https://zfsonlinux.org).
```

* Monitor status with `btrfs replace status ${BTRFSMOUNTPOINT}`.

More sophisticated setups with more than two devices are possible.
Take note that "RAID 1" in `btrfs` means "data is copied on up to two
devices", meaning only up to one device can fail.
You may be interested in `raid1c3` and `raid1c4` modes if you have
three or four storage devices.
BTRFS would probably work better if you were purchasing an entire set
of new storage devices to set up a new node.

`backup` Plugin And Remote NFS Mount
------------------------------------

You can get the `backup` plugin here:
https://github.com/lightningd/plugins/tree/master/backup

The `backup` plugin requires Python 3.

* `cd` into its directory and install requirements.
  * `pip3 install -r requirements.txt`
* Figure out where you will put the backup files.
  * Ideally you have an NFS or other network-based mount on your system,
    into which you will put the backup.
* Stop your Lightning node.
* `/path/to/backup-cli init ${LIGHTNINGDIR} file:///path/to/nfs/mount`.
  This creates an initial copy of the database at the NFS mount.
* Add these settings to your `lightningd` configuration:
  * `important-plugin=/path/to/backup.py`

>   * `backup-destination=file:///path/to/nfs/mount`

Reviewer 3: This is no longer needed, because we now write a `lock` file into the `$LIGHTNINGDIR` in order to be functional right-away, and avoid accidentally changing the backup location.

destination.

Alternately, you *could* put it in another storage device (e.g. USB flash
disk) in the same physical location, but if so, you might want to use the
above `btrfs` RAID-1 instead, as that tends to be more performant
(though using this `backup` plugin does not require you to repartition an
existing hard disk, which might matter to you if you are using an
everyday-use laptop or desktop).

To recover:

* Re-download the `backup` plugin and install Python 3 and the
  requirements of `backup`.
* `/path/to/backup-cli restore file:///path/to/nfs/mount ${LIGHTNINGDIR}`

If your backup destination is a network-mounted filesystem that is in a
remote location, then even loss of all hardware in one location will allow
you to still recover your Lightning funds.

However, if instead you are just replicating the database on another
storage device in a single location, you remain vulnerable to disasters
like fire or computer confiscation.

PostgreSQL Cluster
------------------

`lightningd` may also be compiled with PostgreSQL support.
PostgreSQL is generally faster than SQLITE3, and also supports running a
PostgreSQL cluster to be used by `lightningd`, with automatic replication

> and turnover in case an entire node of the PostgreSQL cluster fails.

Reviewer 3: ```suggestion
and failover in case an entire node of the PostgreSQL cluster fails.
```

Setting this up, however, is more involved.

> By default, `lightningd` compiles without PostgreSQL support.

Reviewer 3: We compile with postgresql by default if `./configure` detect `libpq`, and our releases are configured to ship with postgres enabled.

You then have to compile `lightningd` with PostgreSQL support.

* Clone or untar a new source tree for `lightning` and `cd` into it.
  * You *could* just use `make clean` on an existing one, but for the
    avoidance of doubt (and potential bugs in our `Makefile` cleanup rules),
    just create a fresh source tree.
* `./configure`
  * Add any options to `configure` that you normally use as well.
* Double-check the `config.vars` file contains `HAVE_POSTGRES=1`.
  * `grep 'HAVE_POSTGRES' config.vars`
* `make`
* If you install `lightningd`, `sudo make install`.

If you were not using PostgreSQL before but have compiled and used
`lightningd` on your system, the resulting `lightningd` will still
continue supporting and using your current SQLITE3 database;
it just gains the option to use a PostgreSQL database as well.

To set up a cluster for a brand new node, follow this (external)
[guide by @gabridome][gabridomeguide].

[gabridomeguide]: https://github.com/gabridome/docs/blob/master/c-lightning_with_postgresql_reliability.md

The above guide assumes you are setting up a new node from scratch.
It is also specific to PostgreSQL 12, and setting up for other versions

> **will** have differences; read the PostgreSQL manuals linked above.

Reviewer 3: I'm somehow missing the `--wallet-dsn=postgres://...` pointer to switch over to the postgres DB.

Author: It is `--wallet`, and it is mentioned in the gabridome guide, but will add a random reference to it here.

If you want to continue a node that started using an SQLITE3 database,
use the [tool by @fiatjaf][fiatjafmcldsp] to convert your SQLITE3 database

> to a PostgreSQL database.

Reviewer 5: Hmmm .. Are you confident with it ? Even @fiatjaf [does not seem to be](https://github.com/fiatjaf/mcldsp#beware-this-may-destroy-your-life-and-make-you-lose-all-your-money-use-at-your-own-risk).  If we are confident we are able to provide such a **crucial** tool it shoudl probably be part of the main repo...

Author: No, have never used it, will remove the link instead.

in-channel funds, so this is still better than nothing.

This backup method is undesirable, since it cannot recover the following
channels:

* Channels with peers that do not support `option_dataloss_protect`.
  * Most nodes on the network already support `option_dataloss_protect`
    as of November 2020.
  * If the peer does not support `option_dataloss_protect`, then the entire
    channel funds will be revoked by the peer.
  * Peers can *claim* to honestly support this, but later steal funds
    from you by giving obsolete state when you recover.
* Channels created *after* the copy was made are not recoverable.
  * Data for those channels does not exist in the backup, so your node
    cannot recover them.

Because of the above, this strategy is discouraged: you *can* potentially
lose all funds in open channels.

However, again, note that a "no backups #reckless" strategy leads to
*definite* loss of funds, so you should still prefer *this* strategy rather
than having *no* backups at all.

Even if you have one of the better options above, you might still want to do
this as a worst-case fallback, as long as you:

> * Attempt to recover using the other backup options below first.

Author: s/below/above/

LOL I did not fix it up after rearranging the text!
